webpack/components/Table/TableHooks.js

```diff
@@ -41,14 +42,143 @@ class ReactConnectedSet extends Set {
       this.delete(id);
     }
   }
+
+  addAll(ids) {
+    ids.forEach(id => super.add(id));
+    this.forceRender();
+  }
 }
 
-const useSet = (initialArry) => {
+export const useSet = (initialArry) => {
   const [, setToggle] = useState(0);
   // needed because mutating a Ref won't cause React to rerender
   const forceRender = () => setToggle(prev => prev + 1);
   const set = useRef(new ReactConnectedSet(initialArry, forceRender));
   return set.current;
 };
 
-export default useSet;
+export const useSelectionSet = (results, metadata, initialArry = [], idColumn = 'id') => {
+  const selectionSet = useSet(initialArry);
+  const ids = results?.map(result => result[idColumn]) ?? [];
+  const areAllRowsOnPageSelected = () => Number(ids?.length) > 0 &&
+                                         ids.every(result => selectionSet.has(result));
+
+  const selectPage = () => selectionSet.addAll(ids);
+  const selectNone = () => selectionSet.clear();
+  const onRowSelect = (isSelected, id) => {
+    if (isSelected) {
+      selectionSet.add(id);
+    } else {
+      selectionSet.delete(id);
+    }
+  };
+
+  const selectedCount = selectionSet.size;
+
+  const isSelected = id => selectionSet.has(id);
+
+  return {
+    onRowSelect,
+    selectedCount,
+    areAllRowsOnPageSelected,
+    selectPage,
+    selectNone,
+    isSelected,
+    selectionSet,
+  };
+};
+
+export const useBulkSelect = (results, metadata, initialArry = [], idColumn = 'id') => {
+  const { selectionSet: inclusionSet, ...selectOptions } =
+                useSelectionSet(results, metadata, initialArry, idColumn);
+  const exclusionSet = useSet([]);
+  const [searchQuery, updateSearchQuery] = useState('');
+  const [isSelectAllChecked, setSelectAll] = useState(false);
```

Comment:
I always try to be wary of storing a value in state that could be computed instead.  Instead of trying to keep this value in sync with all the methods below, could we instead make it a function that looks at `inclusionSet`, `exclusionSet`, and `metadata.total` ?

Reply:
If you want to be clever you could even put it in a `useMemo` :wink:

Reply:
Interesting idea. More than the `isSelectAllChecked` I need the `setSelectAllChecked`. How would I figure out the event where selection was set by the `SelectAllCheckbox` component. This state here gives me a handy way to enforce that.